File: src/DP/WildcardMatching.java
package DP;

import java.util.HashMap;
import java.util.Map;

// https://leetcode.com/problems/wildcard-matching/
// similar with RegularExpressionMatching
public class WildcardMatching {
    Map<String, Boolean> cache = new HashMap<>();

    public boolean isMatch(String s, String p) {
        String key = s.length() + ":" + p.length();
        if (cache.containsKey(key)) {
            return cache.get(key);
        }

        if (p.length() == 0 || p == "*") { // when p exhausts or p is only * that could match anything
            return s.length() == 0;
        }

        if (p.charAt(0) == '*') {
            // match * with the first char of s OR match * with 0 char (skip it)
            boolean isMatch = (s.length() >= 1 && isMatch(s.substring(1), p)) || isMatch(s, p.substring(1));
            cache.put(key, isMatch);
            return isMatch;
        } else if (s.length() > 0 && (p.charAt(0) == s.charAt(0) || p.charAt(0) == '?')) {
            // first char is the same, move both forward
            boolean isMatch = isMatch(s.substring(1), p.substring(1));
            cache.put(key, isMatch);
            return isMatch;
        }
        cache.put(key, false);
        return false;
    }

    public boolean isMatchDP(String s, String p) {
        int m = s.length(), n = p.length();
        char[] ws = s.toCharArray();
        char[] wp = p.toCharArray();
        boolean[][] dp = new boolean[m + 1][n + 1];
        dp[0][0] = true;
        for (int j = 1; j <= n; j++)
            dp[0][j] = dp[0][j - 1] && wp[j - 1] == '*';
        for (int i = 1; i <= m; i++)
            dp[i][0] = false;
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (wp[j - 1] == '?' || ws[i - 1] == wp[j - 1])
                    dp[i][j] = dp[i - 1][j - 1];
                else if (wp[j - 1] == '*')
                    dp[i][j] = dp[i - 1][j] || dp[i][j - 1];
            }
        }
        return dp[m][n];
    }

    public static void main(String[] args) {
        WildcardMatching w = new WildcardMatching();
        String s = "adceb", p = "*a*b";
        System.out.println(w.isMatchDP(s, p));
        System.out.println(w.isMatch(s, p));
    }
}

File: py/essential/WildcardMatching.py
# https://leetcode.com/problems/wildcard-matching/description/


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)  # col is p row is s
        dp = [[False] * (n + 1) for _ in range(m + 1)]

        dp[0][0] = True
        for i in range(1, n + 1):
            if p[i - 1] != "*":
                break
            dp[0][i] = True

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if s[i - 1] == p[j - 1] or p[j - 1] == "?":
                    dp[i][j] = dp[i - 1][j - 1]
                elif p[j - 1] == "*":
                    dp[i][j] = (
                        dp[i - 1][j] or dp[i][j - 1]
                    )  # * is treated as matching empty or matching s[i-1]
        return dp[m][n]


s = Solution()
print(s.isMatch("aa", "a"))
print(s.isMatch("aa", "*"))
print(s.isMatch("cb", "?a"))
print(s.isMatch("cb", "?b"))
print(s.isMatch("adceb", "*a*b"))
